ocaml/testsuite/tests/typing-layouts/erasable_annot.ml

```diff
@@ -392,3 +445,102 @@ external f_4 : M3.t -> M3.t = "%identity" [@@unboxed]
 module M3 : sig type t = private float# end
 external f_4 : M3.t -> M3.t = "%identity" [@@unboxed]
 |}];;
+
+(* Disabling warnings *)
+
+module M4 : sig
+  [@@@warning "-187"]
+  type ('a : immediate) t = Something of 'a
+
+  val f : ('a : immediate). 'a t -> 'a
+end = struct
+  [@@@warning "-187"]
+
+  type ('a : immediate) t = Something of 'a
+
+  let f (Something x) = x
+end;;
+
+[%%expect{|
+module M4 :
+  sig
+    type ('a : immediate) t = Something of 'a
+    val f : ('a : immediate). 'a t -> 'a
+  end
+|}]
+
+module[@warning "-187"] M5 = struct
+  let f (type a : immediate): a -> a = fun x -> x
+end;;
+
+[%%expect{|
+module M5 : sig val f : ('a : immediate). 'a -> 'a end
+|}]
+
+(* Just disabling the warning on the expression level doesn't work
+   if the declaration has a type variable annotation. *)
+
+let[@warning "-187"] fails (type a : immediate): a -> a = fun x -> x
+;;
+
+[%%expect{|
+Line 1, characters 21-26:
+1 | let[@warning "-187"] fails (type a : immediate): a -> a = fun x -> x
+                         ^^^^^
+Warning 187 [incompatible-with-upstream]: Usage of layout immediate/immediate64 in fails
+can't be erased for compatibility with upstream OCaml.
+
+val fails : ('a : immediate). 'a -> 'a = <fun>
+|}]
+
+module type S1 = sig
+  type ('a : immediate) fails = int [@@warning "-187"]
+end;;
+
+[%%expect{|
+Line 2, characters 2-54:
+2 |   type ('a : immediate) fails = int [@@warning "-187"]
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Warning 187 [incompatible-with-upstream]: Usage of layout immediate/immediate64 in fails
+can't be erased for compatibility with upstream OCaml.
+
+module type S1 = sig type ('a : immediate) fails = int end
+|}]
+
+(* Disabling the warning just in the signature isn't sufficient.
+
+   The check here is also ran twice for some reason. *)
+module M6 : sig
+  [@@@warning "-187"]
+  type ('a : immediate) t = 'a * 'a
+end = struct
+  type ('a : immediate) t = 'a * 'a
+end;;
+[%%expect{|
+Line 5, characters 2-35:
+5 |   type ('a : immediate) t = 'a * 'a
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Warning 187 [incompatible-with-upstream]: Usage of layout immediate/immediate64 in t
+can't be erased for compatibility with upstream OCaml.
+
+Line 5, characters 2-35:
```

Comment:
Duplicated warning?

Reply:
Might be worth deduplicating the pile of warnings, which would solve various problems here. Or if you'd rather avoid that, we can take a look at this together. The algorithm to check type declarations is pretty twisted.

Reply:
Yeah, I'd appreciate that.